hazelcast/src/main/java/com/hazelcast/cache/impl/operation/LatencyTrackingCacheLoader.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache.impl.operation;
+
+import com.hazelcast.internal.diagnostics.StoreLatencyPlugin;
+
+import javax.cache.integration.CacheLoader;
+import javax.cache.integration.CacheLoaderException;
+import java.util.Map;
+
+public class LatencyTrackingCacheLoader<K, V> implements CacheLoader<K, V> {
+
+    private static final String KEY = "CacheLoaderLatency";
+
+    private final CacheLoader<K, V> delegate;
+    private final StoreLatencyPlugin.LatencyProbe loadProbe;
+    private final StoreLatencyPlugin.LatencyProbe loadAllProbe;
+
+    public LatencyTrackingCacheLoader(CacheLoader<K, V> delegate, StoreLatencyPlugin plugin, String cacheName) {
```

Comment:
How about if the LatencyTracking* classes don't get the plugin but rather have fields annotated with `@Probe`. The decorators would be instantiated not if the plugin is not-null but rather by configuration. Additionally, the `com.hazelcast.internal.metrics.impl.FieldProbe` class would need to be expanded to support a new kind of field (LatencyProbe). Just a suggestion so the decorators don't need to know about plugins.